modules/web-discovery-project/sources/url-analyzer.es

```diff
@@ -0,0 +1,157 @@
+    prefix: "search?q=",
+  },
+  {
+    type: "search-ya",
+    regexp: /^https:[/][/][^/]*[.]search[.]yahoo[.].*?[#?&;][pq]=[^$&]+/,
+    prefix: "search?q=",
+    queryFinder(parsedUrl) {
+      return parsedUrl.searchParams.get("q") || parsedUrl.searchParams.get("p");
+    },
+  },
+  {
+    type: "search-bii",
+    regexp: /^https:[/][/][^/]*[.]bing[.][^/]+[/]images[/]search[?]q=[^$&]+/,
+    prefix: "images/search?q=",
+  },
+  {
+    type: "search-bi",
+    regexp: /^https:[/][/][^/]*[.]bing[.].*?[#?&;]q=[^$&]+/,
+    prefix: "search?q=",
+  },
+  {
+    type: "search-am",
+    regexp:
+      /^https:[/][/][^/]*[.]amazon[.][^/]+[/](s[?]k=[^$&]+|.*[?&]field-keywords=[^$&]+)/,
+    prefix: "s/?field-keywords=",
+    queryFinder(parsedUrl) {
+      return (
+        parsedUrl.searchParams.get("field-keywords") ||
+        parsedUrl.searchParams.get("k")
+      );
+    },
+  },
+  {
+    type: "amp",
+    regexp:
+      /^https:[/][/][^/]*[.]amazon[.][^/]+[/]([/]dp[/]|[/]gp[/]product[/])/,
+    queryFinder(parsedUrl) {
+      return parsedUrl.searchParams.get("keywords");
+    },
+  },
+  {
+    type: "search-dd",
+    regexp:
+      /^https:[/][/]duckduckgo.com[/](?:html$|.*[?&]q=[^&]+.*&ia=web|[?]q=[^&]+$)/,
+    prefix: "?q=",
+  },
+  {
+    type: "li",
+    regexp: /^https:[/][/][^/]*linkedin[.][^/]+[/]pub[/]dir+/,
+  },
+];
+const SEARCH_ENGINE_TYPES = new Set([
+  "search-goi",
+  "search-gov",
+  "search-go",
+  "search-ya",
+  "search-bii",
+  "search-bi",
+  "search-dd",
+]);
+
+export default class UrlAnalyzer {
+  constructor(patterns) {
+    this.patterns = patterns;
+    this._urlPatterns = URL_PATTERNS;
+  }
+
+  parseLinks(url) {
+    for (const {
+      type,
+      regexp,
+      queryFinder = (parsedUrl) => parsedUrl.searchParams.get("q"),
+    } of this._urlPatterns) {
+      if (regexp.test(url)) {
+        // Workaround for an encoding issue (source: https://stackoverflow.com/a/24417399/783510).
+        // Reason: we want to preserve the original search term. In other words, searches
+        // for "abc def" and "abc+def" should be distinguishable. That is why we need to
+        // avoid the ambigious '+' character and use explicit white space encoding.
+        const url_ = url.replaceAll("+", "%20");
+        const parsedUrl = parse(url_);
+
+        const query = queryFinder(parsedUrl);
+        if (!query) {
+          return { found: false };
+        }
+        if (!this.patterns.typeExists(type)) {
+          logger.debug(
+            "Matching rule for",
+            url,
+            "skipped (no matching server side rules exist)",
+          );
+          return { found: false };
+        }
+        return { found: true, type, query };
+      }
+    }
+
+    return { found: false };
+  }
+
+  isSearchEngineUrl(url) {
+    const { found, type } = this.parseLinks(url);
+    if (!found) return false;
+    return SEARCH_ENGINE_TYPES.has(type);
+  }
+
+  tryExtractBraveSerpQuery(url) {
```

Comment:
could you remind me why we monitor Brave Search pages?

Reply:
How do we transmit this event? Is it included in a "query" or "page" message, or something else?